config/v1/types_infrastructure.go

```diff
@@ -51,28 +56,47 @@ type InfrastructureStatus struct {
 type PlatformType string
 
 const (
-	// AWSPlatform represents Amazon Web Services infrastructure.
-	AWSPlatform PlatformType = "AWS"
+	// AWSPlatformType represents Amazon Web Services infrastructure.
+	AWSPlatformType PlatformType = "AWS"
 
-	// AzurePlatform represents Microsoft Azure infrastructure.
-	AzurePlatform PlatformType = "Azure"
+	// AzurePlatformType represents Microsoft Azure infrastructure.
+	AzurePlatformType PlatformType = "Azure"
 
-	// GCPPlatform represents Google Cloud Platform infrastructure.
-	GCPPlatform PlatformType = "GCP"
+	// GCPPlatformType represents Google Cloud Platform infrastructure.
+	GCPPlatformType PlatformType = "GCP"
 
-	// LibvirtPlatform represents libvirt infrastructure.
-	LibvirtPlatform PlatformType = "Libvirt"
+	// LibvirtPlatformType represents libvirt infrastructure.
+	LibvirtPlatformType PlatformType = "Libvirt"
 
-	// OpenStackPlatform represents OpenStack infrastructure.
-	OpenStackPlatform PlatformType = "OpenStack"
+	// OpenStackPlatformType represents OpenStack infrastructure.
+	OpenStackPlatformType PlatformType = "OpenStack"
 
-	// NonePlatform means there is no infrastructure provider.
-	NonePlatform PlatformType = "None"
+	// NonePlatformType means there is no infrastructure provider.
+	NonePlatformType PlatformType = "None"
 
-	// VSpherePlatform represents VMWare vSphere infrastructure.
-	VSpherePlatform PlatformType = "VSphere"
+	// VSpherePlatformType represents VMWare vSphere infrastructure.
+	VSpherePlatformType PlatformType = "VSphere"
 )
 
+// PlatformSpec contains settings that are specific to every underlying infrastructure
+// provider supported by these specs. Since these are used at spec-level, it is a valid
+// scenario to have attributes in more than one platform provider spec.
+type PlatformSpec struct {
+	// AWSPlatform contains settings specific to the Amazon Web Services infrastructure provider.
+	// +optional
+	AWSPlatform AWSPlatformSpec `json:"awsPlatform,omitempty"`
+}
+
+// AWSPlatformSpec specify settings specific to the Amazon Web Services infrastructure provider.
+type AWSPlatformSpec struct {
```

Comment:
Generally you keep fields editable by users in `spec`. This looks like set once by installer and only consumed by others.

Also infraID is relevant for libvirt and other providers too.

Reply:
> Generally you keep fields editable by users in `spec`...

Users should be able to adjust tags, at least for new resources, as long as they preserve the teardown tags.  And ideally (longer term), even those should probably be modifiable via Kubernetes' reconciliation approach.  For now, I'm fine with `Tags` in either or both of the spec and status,

> Also infraID is relevant for libvirt and other providers too.

Ah, right.  Sorry for leading you down the wrong path here, @fabianofranz.

Reply:
> Also infraID is relevant for libvirt and other providers too.

If the `infraID` is valid for every single provider then it could be moved out of the platform specific specs and go directly in `PlatformSpec`, I'd say. Otherwise we'll probably have to duplicate it to the platforms where it applies. WDYT?

And what about `tags`? Is it just for AWS or makes sense in other (maybe all) providers too?

Reply:
> If the `infraID` is valid for every single provider then it could be moved out of the platform specific specs and go directly in `PlatformSpec`, I'd say. 

Abd it's read-only, so move it to `InfrastructureStatus.InfraID`.

Reply:
> And what about `tags`? Is it just for AWS or makes sense in other (maybe all) providers too?

I imagine many, but not all, other platforms will support tags.  But in the install-config it's [an AWS-specific property][1], so I'm fine with where you have it.

[1]: https://github.com/openshift/installer/blob/v0.13.1/pkg/types/aws/platform.go#L9-L11

Reply:
Can we start with these being part of status only, where they will be set once by installer for now.

Reply:
These properties do not feel right in status to me. Looking through: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status 

"The specification is a complete description of the desired state, including configuration settings provided by the user, default values expanded by the system, and properties initialized or otherwise changed after creation by other ecosystem components (e.g., schedulers, auto-scalers), and is persisted in stable storage with the API object."

I would consider this a property initialized by another ecosystem component, i.e. the installer. Spec fields can be made immutable, though that does require a webhook for CRDs I believe, so beyond the scope of this PR and future enhancement, but not allowing a user to edit does not exclude spec per se.

It doesn't seem to fit the definition for status, this is not observed state of the infrastructure, it's part of the specification of the infrastructure.